src/solver/cosineembeddingloss/bwd_unreduced_2d_cosineembeddingloss.cpp

```diff
@@ -0,0 +1,249 @@
+                        (uint64_t)parallelism_size,
+                        (uint64_t)1,
+                        false);
+        if(handle_.IsProfilingEnabled())
+            elapsed += handle_.GetKernelTime();
+
+        auto kernel = handle_.Run(kernels[kernel_cnt++]);
+        kernel(
+            work_b, work_a, (uint64_t)output_numel, (uint64_t)parallelism_size, (uint64_t)1, false);
+
+        if(handle_.IsProfilingEnabled())
+            elapsed += handle_.GetKernelTime();
+    }
+    else
+    {
+        auto kernel = handle_.Run(kernels[kernel_cnt++]);
+        kernel(work_a, work_b, (uint64_t)output_numel, (uint64_t)reduce_size, (uint64_t)1, false);
+        if(handle_.IsProfilingEnabled())
+            elapsed += handle_.GetKernelTime();
+        std::swap(work_a, work_b);
+    }
+}
+
+bool CosineEmbeddingLossUnreducedBackward2d::IsApplicable(
+    const ExecutionContext&,
+    const miopen::cosineembeddingloss::BwdUnreducedProblemDescription& problem) const
+{
+    if(!problem.IsValidLength())
+        return false;
+
+    return true;
+}
+
+ConvSolution CosineEmbeddingLossUnreducedBackward2d::GetSolution(
+    const ExecutionContext& context,
+    const miopen::cosineembeddingloss::BwdUnreducedProblemDescription& problem) const
+{
+    auto result       = ConvSolution{miopenStatusSuccess};
+    auto input_dtype  = miopen::GetDataType(problem.GetInput1Desc().GetType());
+    auto output_dtype = miopen::GetDataType(problem.GetInput1GradDesc().GetType());
+
+    {
+        auto dtype         = problem.GetInput1GradDesc().GetType();
+        size_t Input_total = problem.GetInputTotal();
+
+        const auto build_params = KernelBuildParameters{
+            {"MIOPEN_USE_FP16", static_cast<int>(dtype == miopenHalf)},
+            {"MIOPEN_USE_FP32", static_cast<int>(dtype == miopenFloat)},
+            {"MIOPEN_USE_FP64", static_cast<int>(dtype == miopenDouble)},
+            {"MIOPEN_USE_BFP16", static_cast<int>(dtype == miopenBFloat16)},
+            {"INPUT_TYPE", input_dtype == "bfloat16" ? "ushort" : input_dtype},
+            {"OUTPUT_TYPE", output_dtype == "bfloat16" ? "ushort" : output_dtype},
+            {"D_TYPE", output_dtype == "bfloat16" ? "ushort" : output_dtype},
+        };
+
+        ConstructNormParamsKernelsBwd(context, problem, result, build_params);
+
+        result.construction_params.push_back(
+            make_hip_kernel({LOCAL_SIZE_UNREDUCED_BWD},
+                            {Input_total},
+                            "MIOpenCosineEmbeddingLoss.cpp",
+                            "CosineEmbeddingLossUnreducedBackward2d",
+                            build_params));
+    }
+
+    result.invoker_factory = [](const std::vector<Kernel>& kernels) {
+        return [=](const Handle& handle_, const AnyInvokeParams& raw_params) {
+            decltype(auto) params =
+                raw_params.CastTo<miopen::cosineembeddingloss::BwdInvokeParams>();
+
+            float elapsed = 0.0f;
```

Comment:
If multiple kernels are used, the total op time must be calculated by subtracting the start time of the first kernel from the end time of the last kernel, rather than the sum of the execution times of each kernel. Please refer to src/solver/loss/forward_hinge_embedding_loss.cpp at https://github.com/moreh-dev/MIOpen/pull/20 and modify it.

Reply:
I'm still a bit confused About how we measure kernel duration in my PR: [#20](https://github.com/moreh-dev/MIOpen/pull/20). Can you confirm it, please.
<img width="794" alt="image" src="https://github.com/moreh-dev/MIOpen/assets/52347285/fb5ad643-c4ab-4962-a22b-a53e35fd510b">

Reply:
The comment could not be confirmed because it was resolved. As you said, you can use rocm_op_avg. Are you having any problems?

Reply:
To further explain why you need to measure the entire time, not just the kernel time, the GPU kernel takes time to execute differently from calling a function on the CPU. Therefore, a gap occurs in the section between kernel executions. To reduce this, we sometimes fuse multiple kernels into one.

Reply:
Actually, no. I just wonder if we could use `rocm_op_avg` as in other kernels we use `rocm_kernel_avg`.

Reply:
Of course, the performance comparison result should also be rocm_op_avg, and the result table should specify op total time, not kernel time.

Reply:
I've updated the measurement and result.

Reply:
Are there any changes from the previous measurement method?
There are only some moDNN performance results in this table. Did you encounter any problems?

Reply:
> If multiple kernels are used, the total op time must be calculated by subtracting the start time of the first kernel from the end time of the last kernel, rather than the sum of the execution times of each kernel. Please refer to src/solver/loss/forward_hinge_embedding_loss.cpp at #20 and modify it.

I changed the code as you said here following the [#20](https://github.com/moreh-dev/MIOpen/pull/20) code in commit [86fe6c7](https://github.com/moreh-dev/MIOpen/pull/22/commits/86fe6c735ab78d120b7b8b54e412552b3d39bd3f)

Reply:
Do you mean you also need to add a comparison of Modnn and ROCm to the result table?

Reply:
There are two questions about the difference before and after changing the measurement method and the performance difference between the method used by the existing moDNN and the current method. You can simply comment or create a table.

Reply:
Of course, it is not simply about better performance, but an explanation is needed as to why it is good and in what cases it is good.

Reply:
I've updated the result table. I understand why we should compare the method used by the existing moDNN and the current method, but I don't know why we should compare the difference before and after changing the measurement method. Ain't we supposed to not use the sum of kernel times in my approach? I deleted results of the previous measurement btw.

Reply:
If the op time was measured, the time would probably have increased, but I asked because I was curious about the difference compared to the previous measurement. This is not strictly necessary.

Reply:
Based on your research, performance varies depending on D. Please make two solvers: moDNN method and your new method. Also, check the D value in isapplicable and make sure that only one of the two solvers is selected.(To explain in more detail, Depending on whether isapplicable is passed, the moDNN method is selected if D is less than a certain value, and your method is selected if D is greater than a certain value.)

Reply:
I think that the effect of the kernel gap is not significant in the current op.

Reply:
I've added non sum solvers, please take a look

Reply:
Yes

Reply:
@seungmanhan I have tried your method benchmark for SoftMarginLoss in [this commit](https://github.com/moreh-dev/MIOpen/commit/5ca90a77d95efe4408ad4f2c07f99cb0009b690b). For forward reduced version, which used 2 kernels I encountered a significant decrease in benchmark time. I understand that benchmark with this method the benchmark time will be greater than the method that calculate total kernel time, but it decreased to a number that is not normal. So I also tried to use your method for forward unreduced version, please check code in my commit above. The unreduced forward version only used 1 kernel, so it expected to have similar time when benchmarking with both 2 methods. However, it turned out that benchmark the total kernel time method produced a running time less than your method. Please see benchmark result [here](https://docs.google.com/spreadsheets/d/1cu39WGyEdLY4XBAJ-JFXXxDc3lEg2OaEBKeJQ-LrwmY/edit?usp=sharing) in sheet "compare 2 benchmark methods", column rocm/miopen(kernel) is result that used total kernel time method, column 'rocm_op/miopen_op' is result that used your benchmark method. So I wonder if your benchmark method is reliable. Or maybe the reason is my code or I missed something here. Can you take a look at it?

Reply:
Also, in your sum PR in 2023 [here](https://github.com/moreh-dev/MIOpen/commit/9ec5dc0eb31a3456ab0f559649c77b9332bcaad4#diff-509a445a5b37e34b16d382a095f1828a375119107fdf97f5b5f95ce93857e305) I saw that you didn't use this benchmark method even though it used 2 kernels.

Reply:
The method of directly measuring start and stop has never been analyzed in such detail. Thanks for your analysis.
And compared to the existing method of measuring the kernel start and end time, the method of directly measuring start and stop may be inaccurate for a single kernel (since it is not the actual kernel start and end time).
Therefore, when using two or more kernels, it is recommended to use a method of directly measuring start and stop. (If only two kernels are used, the kernel gap is small, so 3 may be a better number.)
In the case of sum, the method of measuring start and stop was not used because it was a PR created before this method was discovered.

Reply:
@seungmanhan  I have investigated more about the reason why the benchmark time decreased to a number that is not normal. To do this, I print elapsed time for each iteration when running driver. It turned out that there are some iterations that have benchmark time very high to other iter, which affect the overall average time. Below image is result when I run driver for my op SoftMarginLoss, the benchmark time has been sorted. You can see that the largest running time is very different to other iter. I asked other people in our team and they also faced similar problem when benchmarking with this method (the performance decrease very significant), so I think this problem happened because of this benchmark method, not only specific to my operation. And I think this is very serious problem because it decreased performance to a level that from an operation that have performance better than ROCm, it became to an op that worse than ROCm. Because I cannot figured out the root problem of this benchmark method, I came up with a temporary solution to correctly benchmark with this method: we will get median value of iter, I have tested it and it worked seamlessly. How do you think about it?
![image](https://github.com/moreh-dev/MIOpen/assets/33805404/53360ceb-c307-49f1-8f77-ecf08598d997)

Reply:
And for above comment, I am benchmark reduced forward version, which use >= 2 kernels. For case with 1 kernel, I agree with you that we should still use total kernel time method to have correct benchmark result

Reply:
I also had the same problem where performance was suddenly severely reduced in certain cases, and I think the reasons for this could be varied. So I re-tested just for that case and removed any stray values.
Using the median may also be an option. But I think a better way would be to remove only the values ​​that stick out. However, this standard is difficult to set.
You are now a contributor to MIOpen. Please create an issue about this and share it with other people at MIOpen.

Reply:
Actually the severly reduced in certain cases is not the main problem I want to discuss here. The problem is that benchmark with this method reduce performance to a number that is not normal, and also the running time among iterations fluctuate, not consistency like the total kernel time method. I spent several days to investigate this situation and I have just found the root problem. 
First, in invoker_factory, we use handle_.Run to return KernelInvoke, which then we can use this to launch kernel. I looked inside handle_.Run to see what it did:
![image](https://github.com/moreh-dev/MIOpen/assets/33805404/4d78cb16-1426-4932-b4a3-388f0aba6b5c)
When we enable profiling mode (which is our case), it will create HIPOCKernelInvoke with elapsed_time_handler() as a callback function. This callback function will call hipEventElapsedTime, so basically when you run any kernels by using KernelInvoke created by handle_.Run (which all of us are using now), it will auto create start, stop event and call hipEventElapsedTime to measure running time of this kernel. This explained why for operation that only use 1 kernel, calling only kernel = handle_.Run then kernel(...) we can still get profiling result.
That gave me a question: when using  your benchmark  method, you also create new HipEventPtr start, stop and hipEventElapsedTime. So there will be many HipEventPtr in this cases, which can make the overhead time for profiling too large, or they conflict with each other in some ways. I'm not sure about it, but when I tried to disable profiling mode so when calling handle_.Run it will not automatically profile benchmark kernel, then use your benchmark method, the result is more stable, it didn't decreased significantly like previous. It still measured kernel gap between kernels. Below is my code for the method I described.
<img width="655" alt="image" src="https://github.com/moreh-dev/MIOpen/assets/33805404/61bfcd97-d2da-4460-90a9-1128586ed220">

Reply:
Your code makes more sense. When I first used this method, I missed the part about turning off existing profiles in the constructor and destructor. Please share related information with other Vietnamese people through Slack.

Reply:
I have shared it through Slack. Thank you for your review 🤓

Reply:
However, with this benchmark method, the problem " some iterations that have benchmark time very high to other iter" I mentioned in [this comment](https://github.com/moreh-dev/MIOpen/pull/22#discussion_r1628822368) still existed. I don't understand why it didn't happen with total kernel time benchmark method. Maybe I need to create an issue to MIOpen. For now, there are 2 solutions to temporary fix this problem:
1, run only 10 iterations for each test when benchmark. The above problem seems to only happen when I ran >= 50 iters.
2, use median method in the mentioned comment above.

Reply:
I agree to create an issue on MIOpen. I will discuss it with ryu.